services/datamanager/builtin/builtin.go

```diff
@@ -539,30 +544,33 @@ func (svc *builtIn) uploadData(cancelCtx context.Context, intervalMins float64)
 			case <-svc.syncTicker.C:
 				svc.lock.Lock()
 				if svc.syncer != nil {
-					svc.sync(cancelCtx)
+					// If selective sync is not enabled (false), we default to not sync until the trigger
+					// condition has been checked. Otherwise, the default behavior is to sync.
+					shouldSync := !svc.selectiveSyncEnabled
+					// If selective sync is enabled and the sensor has been properly initialized,
+					// try to get the reading from the selective sensor that indicates whether to sync
+					if svc.syncSensor != nil && svc.selectiveSyncEnabled {
+						shouldSync = readyToSync(cancelCtx, svc.syncSensor, svc.logger)
+					}
+					if shouldSync {
+						svc.sync()
+					}
 				}
 				svc.lock.Unlock()
 			}
 		}
 	})
 }
 
-func (svc *builtIn) sync(ctx context.Context) {
+func (svc *builtIn) sync() {
 	svc.flushCollectors()
```

Comment:
I realize this was a bug before — we should only flush collectors if we are about to sync (and `shouldSync` remains true)

Reply:
I'm not sure that I follow, can you expand here? If we've reached here, aren't we about to sync?

Reply:
Not applicable anymore, but previously when we were checking whether to sync within this method, if we didn't, we shouldn't have flushed the collectors.

Reply:
Ah yes! That + design reasons was why I ended up moving the check in this PR